### ISSUE TITLE
How to implement multiple time level storage in FOODiE integrands?

### ISSUE THREAD

**Issue by milancurcic** [Fortran-FOSS-Programmers/FOODIE#12]
With hope of including some multilevel integrators into FOODiE, such as leapfrog or Adams-Bashforth, the integrand type needs to have a capability of storing multiple levels. Currently, the `integrand` type is defined as:

```
type, abstract :: integrand

  private
  real(R_P), dimension(:), allocatable :: state
  ...
endtype integrand
```

where the dimension of `state` represents exactly that - the number of dimensions, that is number of equations in the system. For example, in case of Lorenz equations, the size of `state` is 3 (see https://github.com/Fortran-FOSS-Programmers/FOODiE/blob/master/src/tests/lorenz/lorenz.f90), and in case of oscillation equation, it is 2 (see https://github.com/Fortran-FOSS-Programmers/FOODiE/blob/master/src/tests/oscillation/oscillation.f90).

In the current implementation of integrand type, its state is stored only at one time level, the present one (say, n). However, in order to implement multi-level schemes, it will need to be able to store states at multiple time levels - n, n-1, n-2... etc., and perhaps even n+1, n+2...etc. for implicit integrators.

I suggest that the time storage be implemented as a second rank in the array of states:

```
type, abstract :: integrand

  private
  real(R_P), dimension(:,:), allocatable :: state
  ...
endtype integrand
```

The first index would still reference the equation of the system. The second index would now reference the time level. The time derivative function, for example in the case of oscillation equations:

```
  pure function dOscillation_dt(self) result(dState_dt)
  !---------------------------------------------------------------------------------------------------------------------------------
  !< Time derivative of Oscillation field.
  !---------------------------------------------------------------------------------------------------------------------------------
  class(oscillation), intent(IN) :: self      !< Oscillation field.
  class(integrand),  allocatable :: dState_dt !< Oscillation field time derivative.
  type(oscillation), allocatable :: delta     !< Delta state used as temporary variables.
  !---------------------------------------------------------------------------------------------------------------------------------

  !---------------------------------------------------------------------------------------------------------------------------------
  ! preparing temporary delta
  allocate(delta)
  allocate(delta%state(size(self%state)))
  ! Oscillation equations
  delta%state(1) = -self%f * self%state(2)
  delta%state(2) =  self%f * self%state(1)
  ! hold Oscillation parameters constant over time
  delta%f = 0.
  call move_alloc (delta, dState_dt)
  return
  !---------------------------------------------------------------------------------------------------------------------------------
  endfunction dOscillation_dt
```

would need to be re-implemented to accept the time level as an argument, something like this:

```
  pure function dOscillation_dt(self,n) result(dState_dt)
  !---------------------------------------------------------------------------------------------------------------------------------
  !< Time derivative of Oscillation field.
  !---------------------------------------------------------------------------------------------------------------------------------
  class(oscillation), intent(IN) :: self      !< Oscillation field.
  integer,intent(IN) :: n !< Time level
  class(integrand),  allocatable :: dState_dt !< Oscillation field time derivative.
  type(oscillation), allocatable :: delta     !< Delta state used as temporary variables.
  !---------------------------------------------------------------------------------------------------------------------------------

  !---------------------------------------------------------------------------------------------------------------------------------
  ! preparing temporary delta
  allocate(delta)
  allocate(delta%state(size(self%state)))
  ! Oscillation equations
  delta%state(1,n) = -self%f * self%state(2,n)
  delta%state(2,n) =  self%f * self%state(1,n)
  ! hold Oscillation parameters constant over time
  delta%f = 0.
  call move_alloc (delta, dState_dt)
  return
  !---------------------------------------------------------------------------------------------------------------------------------
  endfunction dOscillation_dt
```

This issue aims to open discussion about the most sensible way to implement this. Let me know if you have any ideas. I will try to implement some multi-level logic and integrators next week.

**Comment by szaghi** [comment]
@milancurcic 
Hi Milan, I am still far from my workstation, thus I cannot check the details, but the abstract type `type_jntegrand` does not (currently) prescribes any member, thus the `state` member can be of any rank (or it can also be absent). The current definition of type integrand prescribes only some type bound procedures (that eventually operate on the integrand members as the `state` could be):

``` fortran
type, abstract :: integrand
  !< Abstract type for building FOODiE ODE integrators.
  contains
    ! public deferred procedures that concrete interpolators must implement
    procedure(time_derivative),      pass(self), deferred, public :: t                            !< Time derivative, residuals.
    procedure(symmetric_operator),   pass(lhs),  deferred, public :: integrand_multiply_integrand !< Integrand * integrand operator.
    procedure(integrand_op_real),    pass(lhs),  deferred, public :: integrand_multiply_real      !< Integrand * real operator.
    procedure(real_op_integrand),    pass(rhs),  deferred, public :: real_multiply_integrand      !< Real * integrand operator.
    procedure(symmetric_operator),   pass(lhs),  deferred, public :: add                          !< Integrand + integrand oprator.
    procedure(assignment_integrand), pass(lhs),  deferred, public :: assign_integrand             !< Integrand = integrand.
    procedure(assignment_real),      pass(lhs),  deferred, public :: assign_real                  !< Integrand = real.
    ! operators overloading
    generic, public :: operator(+) => add                              !< Overloading + operator.
    generic, public :: operator(*) => integrand_multiply_integrand, &
                                      real_multiply_integrand, &
                                      integrand_multiply_real          !< Overloading * operator.
    generic, public :: assignment(=) => assign_integrand, assign_real  !< Overloading = assignament.
endtype integrand
```

The integrators currently operate with only the type bound procedures being _agnostic_ with respect each eventual type integrand member.

This approach works with the current intgrators that include also RK  class that are multi-stages (but not multi-step). The workaround I used for RK integrators is to pass to the integrate method the array of RK stage. However, you could define tbe array as a member of the concrete extension of the type integrand and pass to the integrate method of the multi-step method. On the contrary, defining a _state_ array (of any rank) as a member of the abstract type `type_integrand` is somehow limiting: for example for the Euler equations test that I hope to implement the next week the _state_ array will not a simple real array rather an array of derived type which defines the conservative state (rho, rho_u, rho_E)...

Your question is absolutely crucial and the discussion is open. I vote for a more carefull analysis before prescribes a _state_ member for the abstract type `type_integrand`.

**Comment by milancurcic** [comment]
@szaghi Sorry, I made a mistake and made all this very confusing. In my post, I was writing the `type integrand` definition out of my head, and realized only after reading your reply that indeed, it does not include the state array in it. So, forget about `type integrand`, and I do not suggest adding state array to this.

Instead, consider the `type lorenz` for example, or any other type that extends the integrand type. It has a component that is an allocatable state array with rank 1. My suggestion for storing multiple levels is to implement it as a second dimension, i.e. `real(R_P), dimension(:,:), allocatable :: state`. This would still require modifying the interface for the time derivative function in the `type integrand` definition.

So far, the above has been just ideas in my head and I did not try any coding yet. I will think this through in a bit more detail and try some implementation next week. Let me know if you have any ideas.

**Comment by szaghi** [comment]
@milancurcic 

> Sorry, I made a mistake and made all this very confusing...

Do not worry, as I said your question is fundamental and it is interesting to discuss.

> Instead, consider the type lorenz for example, or any other type that extends the integrand type. It has a component that is an allocatable state array with rank 1. My suggestion for storing multiple levels is to implement it as a second dimension, i.e. real(R_P), dimension(:,:), allocatable :: state. This would still require modifying the interface for the time derivative function in the type integrand definition.

I do not think that modifying the rank of the `state` member of the `type_lorenz` requires to modify also the interface of time derivative function. The abstract interface of the time derivative function is currently:

``` fortran
pure function time_derivative(self) result(dState_dt)
  !---------------------------------------------------------------------------------------------------------------------------------
  !< Time derivative function of integrand class, i.e. the residuals function.
  !---------------------------------------------------------------------------------------------------------------------------------
  import :: integrand
  class(integrand), intent(IN)  :: self      !< Integrand field.
  class(integrand), allocatable :: dState_dt !< Result of the time derivative function of integrand field.
  !---------------------------------------------------------------------------------------------------------------------------------
  endfunction time_derivative
```

Until we embed the `state` member into a `class(integrand)` the time derivative function will operates with the above interface. This is a very abstract and high-level approach: until `field%t` will represent the time derivative in the ODE sense it will depend on only `field`. 

Your concrete extension of the abstract type `type_integrand` can contain whatever you want, but the interface still remain the same, namely only the `type_integrand` should be considered into the time derivative function. At least, this is my current feeling, maybe we found other problems, but the change of the rank of `state` member does not require to change the time derivative API (I guess). On the contrary, the problem could be into the `integrate` procedure of the multi-step(level) class integrators: these require to store some steps (time level) other than the current and question is _how handle these states?_. For the RK class I had to store the RK stages: I have chosen to store theme into an external array not embedded into the `type_lorenz` itself, but I can do. You propose to modify the rank of the `state` member: it is good (clean and efficient) and it requires very few modifications of the tests, but (IMO) it does not require to modify the time derivative API.

In the next week I will come back to office and I will try to implement the integrators you are interested in, I hope that the current API can already support multi-step(level) scheme.

Thank you Milan, see you soon.

**Comment by milancurcic** [comment]
OK, let me try this again. :)

The concrete extension of `type_integrand` also implements the the variant of the `time_derivative` function (BTW, the name "time derivative" I find a bit misleading, since this really just evaluates the user provided function, and does not actually perform a derivative). I suggested that the time derivative implementation, for example `dOscillation_dt`, have an additional `intent(in)` argument n, specifying the time level desired to be integrated:

``` fortran
  pure function dOscillation_dt(self,n) result(dState_dt)
  !---------------------------------------------------------------------------------------------------------------------------------
  !< Time derivative of Oscillation field.
  !---------------------------------------------------------------------------------------------------------------------------------
  class(oscillation), intent(IN) :: self      !< Oscillation field.
  integer,intent(IN) :: n !< Time level
  class(integrand),  allocatable :: dState_dt !< Oscillation field time derivative.
  type(oscillation), allocatable :: delta     !< Delta state used as temporary variables.
  !---------------------------------------------------------------------------------------------------------------------------------

  !---------------------------------------------------------------------------------------------------------------------------------
  ! preparing temporary delta
  allocate(delta)
  allocate(delta%state(size(self%state)))
  ! Oscillation equations
  delta%state(1,n) = -self%f * self%state(2,n)
  delta%state(2,n) =  self%f * self%state(1,n)
  ! hold Oscillation parameters constant over time
  delta%f = 0.
  call move_alloc (delta, dState_dt)
  return
  !---------------------------------------------------------------------------------------------------------------------------------
  endfunction dOscillation_dt
```

so that, in the integrator, for example Adams-Bashforth, you would do something like:

``` fortran
field = field + 0.5*dt*(3*field%t(0)-field%t(-1))
```

I thought that if we add a dummy argument to, say, `dOscillation_dt`, we would also have to add it to `time_derivative`, but please correct me if I'm wrong.

**Comment by szaghi** [comment]
> OK, let me try this again. :)

I like the iterative methods :-)

> the name "time derivative" I find a bit misleading, since this really just evaluates the user provided function, and does not actually perform a derivative

Sure, in practice you are right, but in general `field%t` means just what the math formulation means... a derivative of field with respect the independent variable `t`. However, feel free to correct all my bad nomenclature, my English is very bad.

Seeing your new implementation of the `dOscillation_dt` function I finally understand your point. You are right, if we decide to make the `field%t` procedure a function of time (as it should be, in general we have something like `dy/dt=f(t,y)`) the current abstract interface of the `type_integrand%t` function must be modified.

My initial intention was to leave the `integrate` function blank just to allow such cases: I was planning to pass to the Adams-Bashfort integrate function the time levels array `field(-1:0)` leaving untouched the `field%t` API, but your approach is more clear and more close to the maths.

I suggest to modify  the `type_integrand` abstract interface of `%t` function as following:

``` fortran
abstract interface
  !< Abstract type bound procedures necessary for implementing the class(integrand).
  pure function time_derivative(self, n) result(dState_dt)
  !---------------------------------------------------------------------------------------------------------------------------------
  !< Time derivative function of integrand class, i.e. the residuals function.
  !---------------------------------------------------------------------------------------------------------------------------------
  import :: integrand
  class(integrand), intent(IN)  :: self      !< Integrand field.
  integer, optional, intent(IN)   :: n         !< Time level
  class(integrand), allocatable :: dState_dt !< Result of the time derivative function of integrand field.
  !---------------------------------------------------------------------------------------------------------------------------------
  endfunction time_derivative 
```

Making the time level optional avoids to modify other integrators that are not multi-step/level.

What do you think about?

**Comment by milancurcic** [comment]
@szaghi Excellent! :+1:

**Comment by szaghi** [comment]
@milancurcic see [this](https://github.com/Fortran-FOSS-Programmers/FOODiE/issues/1#issuecomment-136413069).

If, after your review, you think this implementation is good, please close this issue.

See you soon.

**Comment by milancurcic** [comment]
@szaghi Wonderful, thank you! I will review the code and make the tests very soon.

**Comment by milancurcic** [comment]
@szaghi I did a few short tests last night. Your multi-level time storage implementation makes sense and looks correct. However, both AB-2 and AB-3 produce bad results in both Lorenz and Oscillation equations. I am not sure whether the calculation is incorrect, or it is just that the tests were configured outside of the stability region for these schemes - I will do more tests later. So far I can tell I did not find anything wrong with the code itself. Your 1-step AB variant correctly regresses to forward Euler and produces correct solution.

I will keep you posted what I find.

**Comment by szaghi** [comment]
@milancurcic I also see the ab-related issue for the oscillation test. Indeed for the Lorenz I am not completely sure about incorrectness. As I said on gitter I will do accuracy/correctness analysis after the Euler test will be ready. For the moment, I care only if you like the API, in particular:
- how are handled previous times steps, i.e. rank 2 _state_ variable;
- previous times steps cyclic update, i.e. `type_integrand%update_previous_sreps` ;

**Comment by milancurcic** [comment]
@szaghi Sounds good. I agree with the current API regarding multi-level storage. Closing the issue.

**Comment by szaghi** [comment]
@milancurcic :+1:
